File: src/Group.cpp
//////////////////////////////////////////////////////////////////////////
//
//  Copyright (c) 2007, Image Engine Design Inc. All rights reserved.
//
//  Redistribution and use in source and binary forms, with or without
//  modification, are permitted provided that the following conditions are
//  met:
//
//     * Redistributions of source code must retain the above copyright
//       notice, this list of conditions and the following disclaimer.
//
//     * Redistributions in binary form must reproduce the above copyright
//       notice, this list of conditions and the following disclaimer in the
//       documentation and/or other materials provided with the distribution.
//
//     * Neither the name of Image Engine Design nor the names of any
//       other contributors to this software may be used to endorse or
//       promote products derived from this software without specific prior
//       written permission.
//
//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
//  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
//  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
//  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
//  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
//
//////////////////////////////////////////////////////////////////////////

#include "IECore/Group.h"
#include "IECore/Renderer.h"

#include "OpenEXR/ImathBoxAlgo.h"

#include "boost/format.hpp"

using namespace IECore;
using namespace std;
using namespace Imath;

const unsigned int Group::m_ioVersion = 0;
IE_CORE_DEFINEOBJECTTYPEDESCRIPTION( Group );

Group::Group()
	:	m_transform( 0 ), m_parent( 0 )
{
}

Group::~Group()
{
	clearChildren();
}

TransformPtr Group::getTransform()
{
	return m_transform;
}

ConstTransformPtr Group::getTransform() const
{
	return m_transform;
}

void Group::setTransform( TransformPtr transform )
{
	m_transform = transform;
}

Imath::M44f Group::transformMatrix( float time ) const
{
	ConstTransformPtr t = getTransform();
	if( t )
	{
		return t->transform( time );
	}
	return Imath::M44f(); // identity
}

Imath::M44f Group::globalTransformMatrix( float time ) const
{
	if( m_parent )
	{
		return m_parent->globalTransformMatrix( time ) * transformMatrix( time );
	}
	return transformMatrix( time );
}

void Group::addState( StateRenderablePtr state )
{
	if( state->isInstanceOf( Transform::staticTypeId() ) )
	{
		throw Exception( "Transforms cannot be added as state." );
	}
	m_state.insert( state );
}

void Group::removeState( StateRenderablePtr state )
{
	if( m_state.find( state )==m_state.end() )
	{
		throw Exception( "State not present in Group" );
	}
	m_state.erase( state );
}

void Group::clearState()
{
	m_state.clear();
}
		
const Group::StateSet &Group::state() const
{
	return m_state;
}

void Group::addChild( VisibleRenderablePtr child )
{
	GroupPtr gChild = runTimeCast<Group>( child );
	if( gChild )
	{
		GroupPtr gChildParent = gChild->parent();
		if( gChildParent )
		{
			gChildParent->removeChild( gChild );
		}
		gChild->m_parent = this;
	}
	m_children.insert( child );
}

void Group::removeChild( VisibleRenderablePtr child )
{
	ChildSet::iterator it = m_children.find( child );
	
	if( it==m_children.end() )
	{
		throw Exception( "Child is not a member of Group" );
	}
	
	GroupPtr gChild = runTimeCast<Group>( child );
	if( gChild )
	{
		gChild->m_parent = 0;
	}
	
	m_children.erase( it );
}

void Group::clearChildren()
{
	ChildSet::const_iterator it=m_children.begin();
	while( it!=m_children.end() )
	{
		ChildSet::const_iterator next = it; next++;
		removeChild( *it );
		it = next;
	}
}
		
const Group::ChildSet &Group::children() const
{
	return m_children;
}

GroupPtr Group::parent()
{
	return m_parent;
}

ConstGroupPtr Group::parent() const
{
	return m_parent;
}

void Group::copyFrom( ConstObjectPtr other, CopyContext *context )
{
	VisibleRenderable::copyFrom( other, context );
	const Group *tOther = static_cast<const Group *>( other.get() );
	if( tOther->m_transform )
	{
		m_transform = context->copy<Transform>( tOther->m_transform );
	}
	else
	{
		m_transform = 0;
	}
	clearState();
	for( StateSet::const_iterator it=tOther->state().begin(); it!=tOther->state().end(); it++ )
	{
		addState( context->copy<StateRenderable>( *it ) );
	}
	clearChildren();
	for( ChildSet::const_iterator it=tOther->children().begin(); it!=tOther->children().end(); it++ )
	{
		addChild( context->copy<VisibleRenderable>( *it ) );
	}
}

void Group::save( SaveContext *context ) const
{
	VisibleRenderable::save( context );
	IndexedIOInterfacePtr container = context->container( staticTypeName(), m_ioVersion );
	if( m_transform )
	{
		context->save( m_transform, container, "transform" );
	}
	container->mkdir( "state" );
	container->chdir( "state" );
		int i = 0;
		for( StateSet::const_iterator it=state().begin(); it!=state().end(); it++ )
		{
			string name = str( boost::format( "%d" ) % i );
			context->save( *it, container, name );
			i++;
		}
	container->chdir( ".." );
	container->mkdir( "children" );
	container->chdir( "children" );
		i = 0;
		for( ChildSet::const_iterator it = children().begin(); it!=children().end(); it++ )
		{
			string name = str( boost::format( "%d" ) % i );
			context->save( *it, container, name );
			i++;
		}	
	container->chdir( ".." );
}

void Group::load( LoadContextPtr context )
{
	VisibleRenderable::load( context );
	unsigned int v = m_ioVersion;
	IndexedIOInterfacePtr container = context->container( staticTypeName(), v );
	m_transform = 0;
	try
	{
		m_transform = context->load<Transform>( container, "transform" );
	}
	catch( ... )
	{
	}
	clearState();
	container->chdir( "state" );
		IndexedIO::EntryList l = container->ls();
		for( IndexedIO::EntryList::const_iterator it=l.begin(); it!=l.end(); it++ )
		{
			addState( context->load<StateRenderable>( container, it->id() ) );
		}
	container->chdir( ".." );
	clearChildren();
	container->chdir( "children" );
		l = container->ls();
		for( IndexedIO::EntryList::const_iterator it=l.begin(); it!=l.end(); it++ )
		{
			addChild( context->load<VisibleRenderable>( container, it->id() ) );
		}	
	container->chdir( ".." );
}

bool Group::isEqualTo( ConstObjectPtr other ) const
{
	if( !VisibleRenderable::isEqualTo( other ) )
	{
		return false;
	}

	const Group *tOther = static_cast<const Group *>( other.get() );
	
	// check transform
	if( (bool)m_transform != (bool)tOther->m_transform )
	{
		return false;
	}

	if( m_transform && !tOther->m_transform->isEqualTo( m_transform ) )
	{
		return false;
	}
	
	// check state
	if( m_state.size()!=tOther->m_state.size() )
	{
		return false;
	}
	for( StateSet::const_iterator it=m_state.begin(); it!=m_state.end(); it++ )
	{	
		bool found = false;
		for( StateSet::const_iterator it2=tOther->m_state.begin(); it2!=tOther->m_state.end(); it2++ )
		{
			if( (*it)->isEqualTo( *it2 ) )
			{
				found = true;
				break;
			}
		}
		if( !found )
		{
			return false;
		}
	}
	
	// check children
	if( m_children.size()!=tOther->m_children.size() )
	{
		return false;
	}
	for( StateSet::const_iterator it=m_state.begin(); it!=m_state.end(); it++ )
	{	
		bool found = false;
		for( StateSet::const_iterator it2=tOther->m_state.begin(); it2!=tOther->m_state.end(); it2++ )
		{
			if( (*it)->isEqualTo( *it2 ) )
			{
				found = true;
				break;
			}
		}
		if( !found )
		{
			return false;
		}
	}
	
	return true;
}

void Group::memoryUsage( Object::MemoryAccumulator &a ) const
{
	VisibleRenderable::memoryUsage( a );
	a.accumulate( m_transform );
	for( StateSet::const_iterator it=state().begin(); it!=state().end(); it++ )
	{
		a.accumulate( *it );
	}
	for( ChildSet::const_iterator it=children().begin(); it!=children().end(); it++ )
	{
		a.accumulate( *it );
	}
}

void Group::render( RendererPtr renderer )
{
	renderer->attributeBegin();
		if( m_transform )
		{
			m_transform->render( renderer );
		}
		for( StateSet::const_iterator it=state().begin(); it!=state().end(); it++ )
		{
			(*it)->render( renderer );
		}
		for( ChildSet::const_iterator it=children().begin(); it!=children().end(); it++ )
		{
			(*it)->render( renderer );
		}
	renderer->attributeEnd();
}

Imath::Box3f Group::bound() const
{
	Box3f result;
	for( ChildSet::const_iterator it=children().begin(); it!=children().end(); it++ )
	{
		result.extendBy( (*it)->bound() );
	}
	return transform( result, transformMatrix() );
}

File: test/IECoreScene/Group.py
##########################################################################
#
#  Copyright (c) 2007-2010, Image Engine Design Inc. All rights reserved.
#
#  Redistribution and use in source and binary forms, with or without
#  modification, are permitted provided that the following conditions are
#  met:
#
#     * Redistributions of source code must retain the above copyright
#       notice, this list of conditions and the following disclaimer.
#
#     * Redistributions in binary form must reproduce the above copyright
#       notice, this list of conditions and the following disclaimer in the
#       documentation and/or other materials provided with the distribution.
#
#     * Neither the name of Image Engine Design nor the names of any
#       other contributors to this software may be used to endorse or
#       promote products derived from this software without specific prior
#       written permission.
#
#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
#  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
#  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
#  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
#  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
#  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
#  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
#  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
#  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
#  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
#  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#
##########################################################################

import os
import sys
import unittest
import imath
import IECore
import IECoreScene

class TestGroup( unittest.TestCase ) :

	def test( self ) :

		g = IECoreScene.Group()
		self.assertEqual( g.getTransform(), None )
		self.assertEqual( g.transformMatrix(), imath.M44f() )

		g.setTransform( IECoreScene.MatrixTransform( imath.M44f().scale( imath.V3f( 2 ) ) ) )
		self.assertEqual( g.getTransform(), IECoreScene.MatrixTransform( imath.M44f().scale( imath.V3f( 2 ) ) ) )
		self.assertEqual( g.transformMatrix(), imath.M44f().scale( imath.V3f( 2 ) ) )

		self.assertEqual( g.children(), [] )
		self.assertEqual( g.state(), [] )

		# modifying children has no effect on the primitive - children()
		# returns the internal set copied into a list
		g.children().append( IECoreScene.PointsPrimitive( 1 ) )
		self.assertEqual( g.children(), [] )

		g.addChild( IECoreScene.PointsPrimitive( 1 ) )
		self.assertEqual( g.children(), [ IECoreScene.PointsPrimitive( 1 ) ] )

		# modifying state has no effect on the primitive - state()
		# returns the internal set copied into a list
		g.state().append( IECoreScene.AttributeState() )
		self.assertEqual( g.state(), [] )

		g.addState( IECoreScene.AttributeState() )
		self.assertEqual( g.state(), [ IECoreScene.AttributeState() ] )

		self.assertEqual( g, g )

		gg = g.copy()
		self.assertEqual( g, gg )
		self.assertTrue( not gg.children()[0].isSame( g.children()[0] ) )
		self.assertTrue( not gg.state()[0].isSame( g.state()[0] ) )

		IECore.ObjectWriter( g, os.path.join( "test", "group.cob" ) ).write()

		ggg = IECore.ObjectReader( os.path.join( "test", "group.cob" ) ).read()

		self.assertEqual( gg, ggg )
		self.assertTrue( not gg.children()[0].isSame(ggg.children()[0] ) )
		self.assertTrue( not gg.state()[0].isSame(ggg.state()[0] ) )

	def testStateAndChildOrder( self ) :

		# check the state/children don't get reordered when a group is written out to disk
		# and read back in again:
		g = IECoreScene.Group()

		for i in range( 100 ):
			g.addState( IECoreScene.Shader("%d" % i,"ddyup") )

			child = IECoreScene.Group()
			child.blindData()["id"] = IECore.IntData( i )
			g.addChild( child )

		IECore.ObjectWriter( g, os.path.join( "test", "group.cob" ) ).write()

		ggg = IECore.ObjectReader( os.path.join( "test", "group.cob" ) ).read()

		for i in range( 100 ):

			self.assertEqual( g.state()[i].name, ggg.state()[i].name )
			self.assertEqual( g.children()[i].blindData()["id"].value, ggg.children()[i].blindData()["id"].value )



	def testParent( self ) :

		g = IECoreScene.Group()
		g2 = IECoreScene.Group()

		self.assertTrue( g.parent() is None )
		self.assertTrue( g2.parent() is None )

		g.addChild( g2 )

		self.assertTrue( g.parent() is None )
		self.assertTrue( g2.parent().isSame( g ) )

		g.removeChild( g2 )

		self.assertTrue( g.parent() is None )
		self.assertTrue( g2.parent() is None )

		g.addChild( g2 )

		self.assertTrue( g.parent() is None )
		self.assertTrue( g2.parent().isSame( g ) )

		del g
		self.assertTrue( g2.parent() is None )

	def testAttributes( self ) :

		# create a little hierarchy
		g = IECoreScene.Group()
		g2 = IECoreScene.Group()
		g3 = IECoreScene.Group()

		g.addChild( g2 )
		g2.addChild( g3 )

		# define an attribute at the top of the hierarchy
		g.setAttribute( "toptest", IECore.BoolData( False ) )
		self.assertEqual( g.getAttribute( "toptest" ), IECore.BoolData( False ) )

		# change our mind and set it to true:
		g.setAttribute( "toptest", IECore.BoolData( True ) )
		self.assertEqual( g.getAttribute( "toptest" ), IECore.BoolData( True ) )

		# add another attribute
		g.setAttribute( "toptest2", IECore.BoolData( True ) )
		self.assertEqual( g.getAttribute( "toptest2" ), IECore.BoolData( True ) )


		# make sure there's only one AttributeState on the group:
		self.assertEqual( len( g.state() ), 1 )

		# define one in the middle
		g2.setAttribute( "middletest", IECore.BoolData( True ) )

		# override the one at the top
		g2.setAttribute( "toptest", IECore.BoolData( False ) )

		# define one at the bottom
		g3.setAttribute( "bottomtest", IECore.BoolData( False ) )

		self.assertEqual( g.getAttribute( "toptest" ), IECore.BoolData( True ) )
		self.assertEqual( g.getAttribute( "middletest" ), None )
		self.assertEqual( g.getAttribute( "bottomtest" ), None )

		self.assertEqual( g2.getAttribute( "toptest" ), IECore.BoolData( False ) )
		self.assertEqual( g2.getAttribute( "middletest" ), IECore.BoolData( True ) )
		self.assertEqual( g2.getAttribute( "bottomtest" ), None )

		self.assertEqual( g3.getAttribute( "toptest" ), IECore.BoolData( False ) )
		self.assertEqual( g3.getAttribute( "middletest" ), IECore.BoolData( True ) )
		self.assertEqual( g3.getAttribute( "bottomtest" ), IECore.BoolData( False ) )


		# check that the final attribute state is returned by getAttribute:
		g = IECoreScene.Group()
		g.addState( IECoreScene.AttributeState( {"toptest": IECore.BoolData( False ) } ) )
		g.addState( IECoreScene.AttributeState( {"toptest": IECore.BoolData( True ) } ) )

		self.assertEqual( g.getAttribute( "toptest" ), IECore.BoolData( True ) )

		# make sure attributes get added to existing attributeStates:
		g = IECoreScene.Group()
		g.addState( IECoreScene.Shader("yup","ddyup", {}) )
		g.addState( IECoreScene.AttributeState( {"toptest": IECore.BoolData( False ) } ) )
		g.addState( IECoreScene.Shader("yup","yup", {}) )
		g.setAttribute( "blahblah", IECore.BoolData( True ) )

		self.assertEqual( len( g.state() ), 3 )



	def testExceptions( self ) :

		g = IECoreScene.Group()

		self.assertRaises( Exception, g.removeChild, IECoreScene.Group() )
		self.assertRaises( Exception, g.removeState, IECoreScene.AttributeState() )

	def testTransformsNotState( self ) :

		g = IECoreScene.Group()
		self.assertRaises( Exception, g.addState, IECoreScene.MatrixTransform( imath.M44f() ) )

	def testChildOrdering( self ) :

		g = IECoreScene.Group()
		c1 = IECoreScene.PointsPrimitive( 1 )
		c2 = IECoreScene.PointsPrimitive( 2 )
		c3 = IECoreScene.PointsPrimitive( 3 )

		g.addChild( c1 )
		g.addChild( c2 )
		g.addChild( c3 )

		c = g.children()
		self.assertEqual( len( c ), 3 )
		self.assertTrue( c[0].isSame( c1 ) )
		self.assertTrue( c[1].isSame( c2 ) )
		self.assertTrue( c[2].isSame( c3 ) )

	def testStateOrdering( self ) :

		g = IECoreScene.Group()
		a1 = IECoreScene.AttributeState()
		a2 = IECoreScene.AttributeState()
		a3 = IECoreScene.AttributeState()

		g.addState( a1 )
		g.addState( a2 )
		g.addState( a3 )

		s = g.state()
		self.assertEqual( len( s ), 3 )
		self.assertTrue( s[0].isSame( a1 ) )
		self.assertTrue( s[1].isSame( a2 ) )
		self.assertTrue( s[2].isSame( a3 ) )

	def testAddNullState( self ) :

		g = IECoreScene.Group()
		self.assertRaises( Exception, g.addState, None )

	def testAddNullChild( self ) :

		g = IECoreScene.Group()
		self.assertRaises( Exception, g.addChild, None )

	def testNoneRefcount( self ) :

		# exercises a bug whereby we weren't incrementing the reference
		# count for Py_None when returning it to represent a null pointer.
		# this led to "Fatal Python error: deallocating None" type crashes
		g = IECoreScene.Group()
		for i in range( 0, sys.getrefcount( None ) + 100 ) :
			p = g.parent()

	def testMemoryUsage( self ) :

		# this used to crash if the group didn't have a transform
		g = IECoreScene.Group()
		self.assertTrue( g.memoryUsage() > 0 )

	def testHash( self ) :

		g = IECoreScene.Group()
		h = g.hash()

		g.addChild( IECoreScene.SpherePrimitive() )
		self.assertNotEqual( g.hash(), h )
		h = g.hash()

		g.addState( IECoreScene.AttributeState() )
		self.assertNotEqual( g.hash(), h )
		h = g.hash()

		g.setTransform( IECoreScene.MatrixTransform( imath.M44f() ) )
		self.assertNotEqual( g.hash(), h )

	def testGlobalTransform( self ) :

		g = IECoreScene.Group()
		childGroup = IECoreScene.Group()

		g.addChild( childGroup )

		parentTransform = IECore.TransformationMatrixf()
		parentTransform.rotate = imath.Eulerf( 0,3.1415926/2,0 )

		childTransform = IECore.TransformationMatrixf()
		childTransform.translate = imath.V3f( 1, 0, 2 )

		childGroup.setTransform( IECoreScene.MatrixTransform( childTransform.transform ) )
		g.setTransform( IECoreScene.MatrixTransform( parentTransform.transform ) )

		# child group's translation should have been rotated 90 degrees about the y axis:
		s = imath.V3f()
		h = imath.V3f()
		r = imath.V3f()
		childGroupGlobalTranslation = imath.V3f()
		childGroup.globalTransformMatrix().extractSHRT( s, h, r, childGroupGlobalTranslation )
		self.assertAlmostEqual( childGroupGlobalTranslation.x, 2, 4 )
		self.assertAlmostEqual( childGroupGlobalTranslation.y, 0, 4 )
		self.assertAlmostEqual( childGroupGlobalTranslation.z, -1, 4 )



	def tearDown( self ) :

		if os.path.isfile(os.path.join( "test", "group.cob" )):
			os.remove(os.path.join( "test", "group.cob" ))

if __name__ == "__main__":
	unittest.main()
